### PR TITLE
Add distributed test to `RocCurve` metric

### PR DESCRIPTION
**Pull request by sadra-barikbin**
- Add distributed tests for `RocCurve`.
- Override its `compute` method to work in the distributed case.

### CODE REVIEW

File: ignite/contrib/metrics/roc_auc.py [hunk 1]
```diff
@@ -137,15 +139,26 @@ def sigmoid_output_transform(output):
...
             from sklearn.metrics import roc_curve  # noqa: F401
         except ImportError:
             raise ModuleNotFoundError("This contrib module requires scikit-learn to be installed.")
 
         super(RocCurve, self).__init__(
-            roc_auc_curve_compute_fn, output_transform=output_transform, check_compute_fn=check_compute_fn
+            roc_auc_curve_compute_fn,
+            output_transform=output_transform,
+            check_compute_fn=check_compute_fn,
+            device=device,
```
Comment (vfdev-5) [inline ignite/contrib/metrics/roc_auc.py:163]: @sadra-barikbin you have to override `compute` method like in PrecisionRecallCurve: https://github.com/pytorch/ignite/blob/7c8d801b68c2ed5bc0c030e124df6bb4c027197a/ignite/contrib/metrics/precision_recall_curve.py#L87-L118

**Review by vfdev-5 — DISMISSED** [review]
Thanks @sadra-barikbin !

**Comment by vfdev-5** [comment]
@sadra-barikbin somehow the tests are stuck on nightly ... we have to figure out that

**Comment by vfdev-5** [comment]
@sadra-barikbin can you check what happens with nightly tests, why they stuck ?

**Comment by sadra-barikbin** [comment]
The only test failure is because `torch._six` is removed 2 days ago and torchvision updated its code accordingly:
https://github.com/pytorch/vision/pull/7243.

File: tests/ignite/contrib/metrics/test_roc_curve.py [hunk 2]
```diff
@@ -121,3 +132,26 @@ def test_check_compute_fn():
...
+    rank = idist.get_rank()
+    torch.manual_seed(41 + rank)
+    n_batches, batch_size = 5, 10
+    y = torch.randint(0, 2, size=(n_batches * batch_size,))
+    y_pred = torch.rand((n_batches * batch_size,))
+
+    device = "cpu" if idist.device().type == "xla" else idist.device()
+    metric = RocCurve(device=device)
+    for i in range(n_batches):
+        metric.update((y_pred[i * batch_size : (i + 1) * batch_size], y[i * batch_size : (i + 1) * batch_size]))
```
Comment (vfdev-5) [inline tests/ignite/contrib/metrics/test_roc_curve.py:147]: Can you replace this using `Engine` and a proper engine integration. There are similar tests for other contrib metrics to inspire from.

**Comment by vfdev-5** [comment]
> The only test failure is because `torch._six` is removed 2 days ago and torchvision updated its code accordingly: [pytorch/vision#7243](https://github.com/pytorch/vision/pull/7243).

Yes, this test failure is unrelated. Please address this comment https://github.com/pytorch/ignite/pull/2802#discussion_r1108508284 and we'll be good to go

File: ignite/contrib/metrics/roc_auc.py [hunk 3]
```diff
@@ -137,15 +141,57 @@ def sigmoid_output_transform(output):
...
+
+        _prediction_tensor = torch.cat(self._predictions, dim=0)
+        _target_tensor = torch.cat(self._targets, dim=0)
+
+        ws = idist.get_world_size()
+        if ws > 1 and not self._is_reduced:
+            # All gather across all processes
+            _prediction_tensor = cast(torch.Tensor, idist.all_gather(_prediction_tensor))
+            _target_tensor = cast(torch.Tensor, idist.all_gather(_target_tensor))
+        self._is_reduced = True
```
Comment (vfdev-5) [inline ignite/contrib/metrics/roc_auc.py:180]: Maybe we should not set `self._is_reduced` at all here.
```suggestion
        if ws > 1:
            # All gather across all processes
            _prediction_tensor = cast(torch.Tensor, idist.all_gather(_prediction_tensor))
            _target_tensor = cast(torch.Tensor, idist.all_gather(_target_tensor))
```

**Review by vfdev-5 — APPROVED** [review]
Thanks @sadra-barikbin

**Comment by vfdev-5** [comment]
GPU tests are failing : https://app.circleci.com/pipelines/github/pytorch/ignite/2996/workflows/0dd3e544-0b2c-445d-9a04-1f2a0544eb21/jobs/8141/parallel-runs/0/steps/0-107
